pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentTopic.java

```diff
@@ -1423,6 +1439,38 @@ public CompletableFuture<Void> checkReplication() {
             }
         });
 
+        futures.add(checkShadowReplication());
+
+        return FutureUtil.waitForAll(futures);
+    }
+
+    private CompletableFuture<Void> checkShadowReplication() {
+        if (CollectionUtils.isEmpty(shadowTopics)) {
+            return CompletableFuture.completedFuture(null);
+        }
+        List<String> configuredShadowTopics = shadowTopics;
```

Comment:
`shadowTopics` can be updated concurrently. We just need to handle the snapshot of the value.

Reply:
The update of `shadowTopics` will lead to `configuredShadowTopics` update, because they refer to same object.

If I understand correctly, we should write like below, which will copy the `shadowTopics` at the moment:

```
List<String> configuredShadowTopics = new ArrayList<>(shadowTopics);
```

WDYT @Jason918

Reply:
@AnonHxy shadowTopics is volatile and it will refers to new array object in `onUpdate(TopicPolicies policies)`. So we don't need copy there.